### DISCUSSION TITLE
Rename via editor

### DISCUSSION THREAD

**Discussion by innocenzi** [75lb/renamer#55] [Ideas]
I was going to make a library for this, and while looking for a name, I stumbled upon this one. So instead of creating a library, I'll see if my idea fits in this one. :)

The idea is to be able to rename a bunch of files directly in the editor. This came to my mind when I saw a bunch of files and thought, "it'd be cool if I could multicursor-rename those". 

I'm thinking of the same behavior than `git commit` regarding the editing: 

- You use `renamer` in a directory
- Your configured editor opens, with a file name in each line
- You rename them as you want and close the editor
- `renamer` renames each file according to their new names

For instance:

```batch
➜  ll
total 8
drwxr-xr-x 2 innocenzi innocenzi 4096 Jun  9 16:07  .
drwxr-xr-x 4 innocenzi innocenzi 4096 Jun  9 16:07  ..
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 01 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 02 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 03 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 04 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 05 [1080p].mkv'
➜  renamer
hint: Waiting for your editor to close the file...
```

At this moment, the editor is opened with the following content: 

```
[HorribleSubs] Tower of God - 01 [1080p].mkv
[HorribleSubs] Tower of God - 02 [1080p].mkv
[HorribleSubs] Tower of God - 03 [1080p].mkv
[HorribleSubs] Tower of God - 04 [1080p].mkv
[HorribleSubs] Tower of God - 05 [1080p].mkv
```

You rename them with your multi-cursor skills:

![](https://i.imgur.com/9ifGfl3.gif)

And you close your editor. 

```batch
➜  ll
total 8
drwxr-xr-x 2 innocenzi innocenzi 4096 Jun  9 16:07  .
drwxr-xr-x 4 innocenzi innocenzi 4096 Jun  9 16:07  ..
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 01 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 02 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 03 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 04 [1080p].mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:07 '[HorribleSubs] Tower of God - 05 [1080p].mkv'
➜  renamer
5 files renamed.
➜  ll
total 8
drwxr-xr-x 2 innocenzi innocenzi 4096 Jun  9 16:07  .
drwxr-xr-x 4 innocenzi innocenzi 4096 Jun  9 16:07  ..
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:08 'Tower of God - 01.mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:08 'Tower of God - 02.mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:08 'Tower of God - 03.mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:08 'Tower of God - 04.mkv'
-rw-r--r-- 1 innocenzi innocenzi    0 Jun  9 16:08 'Tower of God - 05.mkv'
```

The advantage is to not have to remember a CLI-specific syntax, and to have the power of your editor under the hand. 
The only challenge is to get the editor-thing right. I believe `git commit`'s behavior is perfect, but the implementation may be challenging to get done. 

What do you think?

**Comment by 75lb** [comment]
good idea, nicely presented! In your tool, what happens if the user deletes a line in the editor ? The renaming app would be able to detect a line had been removed but would be confused which file to rename to what.. 

For example, say these are your input files ("before" state): 

```
photo-of-a-cow.jpg
racing-car-action-shot.jpg
untitled-photo.png
```

and in your editor you renamer the files to this ("after" state):

```
image-of-an-Angus-foal.jpg
my-back-garden.png
```

One file has been removed. Should the renaming app rename `racing-car-action-shot.jpg` or `untitled-photo.png` to `my-back-garden.png`?

**Reply by innocenzi** [reply]
I'm thinking of three options: 

- We find a way to map the remaining lines to their original options (seems hard, and probably unreliable)
	- Either we ignore the missing file
	- Or we delete it
- We detect that the number of lines doesn't match, and we ask the user what to do:
	- Either cancel
	- Or agree, but the files after the removed lines will be offset?

---

Alternatively, if the user erases the line (but keep it), like this: 

```
photo-of-a-cow.jpg

untitled-photo.png
```

It would delete the corresponding file (`racing-car-action-shot.jpg`). 

I've a MVP working locally which implements the second option if you're interested, but it was fairly simple to do. For the editor, the implementation is way easier than what I thought, thanks to [this](https://git-scm.com/book/en/v2/Appendix-C%3A-Git-Commands-Setup-and-Config). We basically just start the process and wait until it ends. The hardest part now is the configurability.

---

**EDIT** - Since I had an immediate use for it, I published the tool: https://github.com/innocenzi/rename. Feel free to use the code to implement it in `renamer` :)
